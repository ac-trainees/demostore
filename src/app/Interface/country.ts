export interface CountryI {
  name: string;
  alpha2Code: string;
}
