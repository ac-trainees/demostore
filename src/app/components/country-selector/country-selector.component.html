<button mat-icon-button>
  <span
    color="primary"
    yPosition="before"
    [matMenuTriggerFor]="menu"
    #ClickMatMenuTrigger="matMenuTrigger"
    (click)="resetFilter()"
    class="fi fi-{{ country }}"
  ></span>
</button>

<mat-menu #menu="matMenu">
  <div (click)="$event.stopPropagation()">
    <mat-list>
      <div mat-subheader>Country:</div>
      <br />
      <mat-list-item>
        <mat-form-field appearance="fill">
          <mat-label>Seach</mat-label>
          <input
            matInput
            [(ngModel)]="seachValue"
            (ngModelChange)="filterCountrys()"
          />
        </mat-form-field>
      </mat-list-item>
      <mat-list-item *ngFor="let country of filteredCountrys">
        <button mat-menu-item (click)="setCountry(country.alpha2Code)">
          <span class="fi fi-{{ country.alpha2Code }}"></span>
          {{ country.name }}
        </button>
      </mat-list-item>
    </mat-list>
  </div>
</mat-menu>
