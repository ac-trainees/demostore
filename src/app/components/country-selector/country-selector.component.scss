@import url(~flag-icons/css/flag-icons.css);
