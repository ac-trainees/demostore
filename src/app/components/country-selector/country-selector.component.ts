import { CountryService } from '../../services/country.service';
import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'app-country-selector',
  templateUrl: './country-selector.component.html',
  styleUrls: ['./country-selector.component.scss'],
})
export class CountrySelectorComponent implements OnInit {
  country!: string;
  seachValue!: string;

  allCountrys: { short: string; long: string }[] = [
    { short: 'en', long: 'United States' },
    { short: 'de', long: 'Germany' },
  ];

  filteredCountrys!: { short: string; long: string }[];

  constructor(public countryService: CountryService) {}

  ngOnInit(): void {
    this.country = this.countryService.country;
    this.filteredCountrys = this.allCountrys;
  }

  filterCountrys() {
    this.filteredCountrys = this.allCountrys.filter((element) => {
      return element.long.includes(this.seachValue);
    });
  }

  setCountry(shortCountry: string): void {
    this.country = shortCountry;
    this.countryService.setService(shortCountry);
  }
}
