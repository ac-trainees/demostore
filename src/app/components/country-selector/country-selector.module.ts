import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatMenuModule } from '@angular/material/menu';
import { MatButtonModule } from '@angular/material/button';
import { CountrySelectorComponent } from './country-selector.component';

@NgModule({
  declarations: [CountrySelectorComponent],
  imports: [MatMenuModule, CommonModule, MatButtonModule],
  exports: [CountrySelectorComponent],
})
export class CountrySelectorModule {}
