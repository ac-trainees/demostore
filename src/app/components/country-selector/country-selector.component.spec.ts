import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { MatListModule } from '@angular/material/list';
import { MatMenuModule } from '@angular/material/menu';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { CountryService } from '../../services/country.service';
import { CountrySelectorComponent } from './country-selector.component';

describe('CountrySelectorComponent', () => {
  let component: CountrySelectorComponent;
  let fixture: ComponentFixture<CountrySelectorComponent>;

  let countryServiceStub = {
    country: 'error',
    setService: jest.fn((country: string) => {
      countryServiceStub.country = country;
    }),
  };

  let eventStub: any = { stopPropagation: jest.fn((event: Event) => {}) };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [CountrySelectorComponent],
      imports: [
        MatMenuModule,
        MatButtonModule,
        MatListModule,
        MatFormFieldModule,
        FormsModule,
        MatInputModule,
        BrowserAnimationsModule,
      ],
      providers: [{ provide: CountryService, useValue: countryServiceStub }],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(CountrySelectorComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should set local country AND countryService', () => {
    component.setCountry('de');

    expect(component.country).toBe('de');
    expect(countryServiceStub.country).toBe('de');
  });

  it('should filter the list of countries', () => {
    component.seachValue = 'G';
    component.filterCountrys();
    expect(component.filteredCountrys).toContainEqual({
      alpha2Code: 'de',
      name: 'Germany',
    });
  });

  it('should filter the list of countries', () => {
    component.seachValue = 'U';
    component.filterCountrys();
    expect(component.filteredCountrys).toContainEqual({
      alpha2Code: 'us',
      name: 'United States',
    });
  });

  it('should be called', () => {
    component.stopPropagation(eventStub);
    expect(eventStub.stopPropagation.mock.calls.length).toBe(1);
  });

  it('should reset filter ad run filter', () => {
    component.resetFilter();
    expect(component.seachValue).toBe('');
    expect(component.allCountrys).toHaveLength(2);
  });
});
