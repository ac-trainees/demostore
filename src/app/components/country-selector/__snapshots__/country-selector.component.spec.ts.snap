// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelectorComponent should create 1`] = `
<app-country-selector
  __ngContext__={[Function LRootView]}
  allCountrys={[Function Array]}
  country={[Function String]}
  countryService={[Function Object]}
  filteredCountrys={[Function Array]}
  menuOpen="false"
  trigger={[Function MatMenuTrigger]}
>
  <button
    aria-haspopup="true"
    class="mat-focus-indicator mat-menu-trigger mat-raised-button mat-button-base mat-primary"
    color="primary"
    mat-raised-button=""
    yposition="before"
  >
    <span
      class="mat-button-wrapper"
    >
      <span
        class="fi fi-us"
      >
        ERROR
      </span>
    </span>
    <span
      class="mat-ripple mat-button-ripple"
      matripple=""
    />
    <span
      class="mat-button-focus-overlay"
    />
  </button><mat-menu>
  </mat-menu>
</app-country-selector>
`;
